### ISSUE TITLE
FP SING_SINGLETON_HAS_NONPRIVATE_CONSTRUCTOR for record

### ISSUE THREAD

**Issue by xzel23** [spotbugs/spotbugs#2981]
After #2934 was fixed, most problems with singletons seem to be solved. There still is one left. I don't know if there's an easy way to fix this - if not, I'll add an exclusion rule in my code.

I have the following record class where SpotBugs still reports a false positive SING_SINGLETON_HAS_NONPRIVATE_CONSTRUCTOR:

```
    record RowInfo(int firstRow, int lastRow) {
        private static final RowInfo NONE = new RowInfo(0, -1);

        public static RowInfo none() {
            return NONE;
        }
    }
```

The problem I think is that it's not easy to tell if something should be a singleton or not. And you could implement a singleton as a record, but it is rather usual IMHO. I this example, the static final member is not intended as a singleton instance, but is used as a guard value.

**Comment by JuditKnoll** [comment]
Thanks for reporting this issue. I will take a closer look at it next week.

**Comment by gtoison** [comment]
A record must have a non private constructor, so it can't be a singleton here.
For non record classes it will be difficult to decide if the intent was to write a singleton

**Comment by xzel23** [comment]
@gtoison a private record can have only private constructors. I don't know how relevant that is for "real" code. Could be used as a singleton holder for the enclosing class.

**Comment by gtoison** [comment]
Ah you're right, I got confused with the canonical record constructor
